<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="fionarhuang" />

<meta name="date" content="2021-09-18" />

<title>A toy example of using treeclimbR on multiple features per entity</title>

<script src="site_libs/header-attrs-2.10/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">DS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">A toy example of using treeclimbR on multiple features per entity</h1>
<h4 class="author">fionarhuang</h4>
<h4 class="date">2021-09-18</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-09-18
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>DS/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20210918code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20210918)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20210918code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20210918)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongbaf3539"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> baf3539 </a>
</p>
</div>
<div id="strongRepositoryversionstrongbaf3539" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version baf3539. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/DS_toy.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<pre class="r"><code>suppressPackageStartupMessages({
  library(TreeSummarizedExperiment)
  library(ggtree)
  library(ape)
  library(treeclimbR)
  library(TreeHeatmap)
  library(ggnewscale)
  library(ggplot2)
}) </code></pre>
<p>This a toy example to help understand how treeclimbR works on the <code>DS</code> case where each entity (e.g., cell subpopulation in scRNA-seq data) has multiple features (e.g., genes). The main difference in data structure between the <code>DA</code> and the <code>DS</code> case is in the number of features that need to be taken into account in the FDR control.</p>
<div id="generate-a-toy-dataset" class="section level1">
<h1>Generate a toy dataset</h1>
<div id="tree" class="section level2">
<h2>Tree</h2>
<pre class="r"><code># Generate a random tree
set.seed(1)
n &lt;- 100
tree &lt;- rtree(n)

# Add labels on nodes
tree &lt;- addLabel(tree = tree, 
                 label = c(paste0(&quot;L&quot;, seq_len(n)), paste0(&quot;N&quot;, seq_len(n-1))),
                 on = &quot;all&quot;)</code></pre>
</div>
<div id="count-tables" class="section level2">
<h2>Count tables</h2>
<p>Simulate counts of 10 features for 100 entities in 30 samples. 15 samples in group A, the other 15 samples in group B.</p>
<pre class="r"><code># The proportions of features 
pfun &lt;- function(n, shape1 = 2, shape2 = 3) {
  p &lt;- rbeta(n = n, shape1 = shape1, shape2 = shape2)
  p/sum(p)
}
pmat &lt;- rbind(c(2, 3), c(2, 2), c(3, 5), c(2, 5), c(1, 4),
              c(3, 4), c(3, 2), c(4, 3), c(1, 5), c(1, 3)) 
rownames(pmat) &lt;- paste0(&quot;gene&quot;, 1:10)
plist &lt;- apply(pmat, 1, FUN = function(x){
  pfun(n = 100, shape1 = x[1], shape2 = x[2])
})


# Generate counts of features
## Create samples for two groups, each group having m samples
m &lt;- 15
countList &lt;- lapply(seq_len(ncol(plist)), FUN = function(x){
  xx &lt;- rmultinom(n = m*2, size = 2000, prob = plist[, x])
  rownames(xx) &lt;- paste0(&quot;L&quot;, 1:100)
  colnames(xx) &lt;- paste0(rep(LETTERS[1:2], each = m), c(1:m, 1:m))
  return(xx)
})


# randomly select 5 internal nodes in the tree
intNode &lt;- setdiff(showNode(tree, only.leaf = FALSE),
               showNode(tree, only.leaf = TRUE))
nd &lt;- sample(x = intNode, size = 5, replace = FALSE)
names(nd) &lt;- paste0(&quot;feature&quot;, 1:5)

# For descendant leaves of the selected nodes:
# - Multiply their counts in the first group by a value fc
sigList &lt;- findDescendant(tree = tree, node = nd)
sigList &lt;- lapply(sigList, convertNode, tree = tree)
fc &lt;- 3

for (i in seq_along(sigList)) {
  countList[[i]][sigList[[i]], 1:m] &lt;- countList[[i]][sigList[[i]], 1:m] * fc
}
names(countList) &lt;- paste0(&quot;feature&quot;, 1:10)</code></pre>
</div>
<div id="create-a-treese-object" class="section level2">
<h2>create a treeSE object</h2>
<pre class="r"><code>sampInfo &lt;- data.frame(group = rep(LETTERS[1:2], each = m))
lse &lt;- TreeSummarizedExperiment(assays = countList, rowTree = tree, 
                                colData = sampInfo)
lse</code></pre>
<pre><code>class: TreeSummarizedExperiment 
dim: 100 30 
metadata(0):
assays(10): feature1 feature2 ... feature9 feature10
rownames(100): L1 L2 ... L99 L100
rowData names(0):
colnames(30): A1 A2 ... B14 B15
colData names(1): group
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (100 rows)
rowTree: 1 phylo tree(s) (100 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
<div id="data-aggregation" class="section level2">
<h2>Data aggregation</h2>
<pre class="r"><code>all_node &lt;- showNode(tree = tree, only.leaf = FALSE)
tse &lt;- aggTSE(x = lse, rowLevel = all_node, rowFun = sum, whichAssay = 1:10)
rownames(tse) &lt;- convertNode(tree = rowTree(tse), node = rowLinks(tse)$nodeNum)
tse</code></pre>
<pre><code>class: TreeSummarizedExperiment 
dim: 199 30 
metadata(0):
assays(10): feature1 feature2 ... feature9 feature10
rownames(199): L1 L2 ... N98 N99
rowData names(0):
colnames(30): A1 A2 ... B14 B15
colData names(1): group
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (199 rows)
rowTree: 1 phylo tree(s) (100 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<div id="viz-the-toy-data" class="section level1">
<h1>Viz the toy data</h1>
<p>Only five features (features 1-5) have signals. Each feature has one signal branch with branch 1-5 for features 1-5 correspondingly. The five signal branches are randomly selected and vary in branch sizes. Heatmaps are simulated counts of features in 30 samples that belong to two different groups. Samples are in columns. The first 15 samples belong to the group A, and the rest belong to the group B.</p>
<pre class="r"><code>ggtree(tree, branch.length = &quot;none&quot;) +
  geom_hilight(node = nd[1], fill = &quot;red&quot;, alpha = 0.2, extend = 1.5) +
  geom_hilight(node = nd[2], fill = &quot;blue&quot;, alpha = 0.2, extend = 1.5) +
  geom_hilight(node = nd[3], fill = &quot;orange&quot;, alpha = 0.2, extend = 1.5) +
  geom_hilight(node = nd[4], fill = &quot;yellow&quot;, alpha = 0.2, extend = 1.5) +
  geom_hilight(node = nd[5], fill = &quot;navy&quot;, alpha = 0.2, extend = 1.5) +
  geom_cladelab(node = nd[1], label =  &quot;1&quot;, fill = &quot;red&quot;, fontsize = 3) +
  geom_cladelab(node = nd[2], label =  &quot;2&quot;, fill = &quot;red&quot;, fontsize = 3) +
  geom_cladelab(node = nd[3], label =  &quot;3&quot;, fill = &quot;red&quot;, fontsize = 3) +
  geom_cladelab(node = nd[4], label =  &quot;4&quot;, fill = &quot;red&quot;, fontsize = 3) +
  geom_cladelab(node = nd[5], label =  &quot;5&quot;, fill = &quot;red&quot;, fontsize = 3) +
  geom_point2(aes(subset = node %in% c(nd)), color = &quot;red&quot;) +
  geom_th_heatmap(name = &quot;feature1&quot;, th_data = log10(assays(lse)[[1]]), rel_width = 0.3, gap = 2) + 
  geom_th_title(name = &quot;feature1&quot;, side = &quot;top&quot;, label = &quot;feature1&quot;, size = 2, nudge_y = 1) +
  scale_fill_viridis_c(option = &quot;A&quot;) +
  new_scale_fill() +
  geom_th_heatmap(name = &quot;feature2&quot;, th_data = log10(assays(lse)[[2]]), rel_width = 0.3) + 
  geom_th_title(name = &quot;feature2&quot;, side = &quot;top&quot;, label = &quot;feature2&quot;, size = 2, nudge_y = 1) +
  scale_fill_viridis_c(option = &quot;A&quot;) +
  new_scale_fill() +
  geom_th_heatmap(name = &quot;feature3&quot;, th_data = log10(assays(lse)[[3]]), rel_width = 0.3) + 
  geom_th_title(name = &quot;feature3&quot;, side = &quot;top&quot;, label = &quot;feature3&quot;, size = 2, nudge_y = 1) +
  scale_fill_viridis_c(option = &quot;A&quot;) +
  new_scale_fill() +
  geom_th_heatmap(name = &quot;feature4&quot;, th_data = log10(assays(lse)[[4]]), rel_width = 0.3) + 
  geom_th_title(name = &quot;feature4&quot;, side = &quot;top&quot;, label = &quot;feature4&quot;, size = 2, nudge_y = 1) +
  scale_fill_viridis_c(option = &quot;A&quot;) +
  new_scale_fill() +
  geom_th_heatmap(name = &quot;feature5&quot;, th_data = log10(assays(lse)[[5]]), rel_width = 0.3) + 
  geom_th_title(name = &quot;feature5&quot;, side = &quot;top&quot;, label = &quot;feature5&quot;, size = 2, nudge_y = 1) +
  scale_fill_viridis_c(option = &quot;A&quot;) +
  theme(legend.position = &quot;none&quot;) +
  scale_y_continuous(expand = c(0.1, 0.1))</code></pre>
<p><img src="figure/DS_toy.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Features 6-10 have no signal.</p>
<pre class="r"><code>ggtree(tree, branch.length = &quot;none&quot;) +
  geom_th_heatmap(name = &quot;feature6&quot;, th_data = log10(assays(lse)[[6]]), rel_width = 0.3, gap = 2) + 
  geom_th_title(name = &quot;feature6&quot;, side = &quot;top&quot;, label = &quot;feature6&quot;, size = 2, nudge_y = 1) +
  scale_fill_viridis_c(option = &quot;A&quot;) +
  new_scale_fill() +
  geom_th_heatmap(name = &quot;feature7&quot;, th_data = log10(assays(lse)[[7]]), rel_width = 0.3) + 
  geom_th_title(name = &quot;feature7&quot;, side = &quot;top&quot;, label = &quot;feature7&quot;, size = 2, nudge_y = 1) +
  scale_fill_viridis_c(option = &quot;A&quot;) +
  new_scale_fill() +
  geom_th_heatmap(name = &quot;feature8&quot;, th_data = log10(assays(lse)[[8]]), rel_width = 0.3) + 
  geom_th_title(name = &quot;feature8&quot;, side = &quot;top&quot;, label = &quot;feature8&quot;, size = 2, nudge_y = 1) +
  scale_fill_viridis_c(option = &quot;A&quot;) +
  new_scale_fill() +
  geom_th_heatmap(name = &quot;feature9&quot;, th_data = log10(assays(lse)[[9]]), rel_width = 0.3) + 
  geom_th_title(name = &quot;feature9&quot;, side = &quot;top&quot;, label = &quot;feature9&quot;, size = 2, nudge_y = 1) +
  scale_fill_viridis_c(option = &quot;A&quot;) +
  new_scale_fill() +
  geom_th_heatmap(name = &quot;feature10&quot;, th_data = log10(assays(lse)[[10]]), rel_width = 0.3) + 
  geom_th_title(name = &quot;feature10&quot;, side = &quot;top&quot;, label = &quot;feature10&quot;, size = 2, nudge_y = 1) +
  scale_fill_viridis_c(option = &quot;A&quot;) +
  theme(legend.position = &quot;none&quot;) +
  scale_y_continuous(expand = c(0.1, 0.1))</code></pre>
<p><img src="figure/DS_toy.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="differential-analysis" class="section level1">
<h1>Differential analysis</h1>
<p>The non-parametric method, Wilcoxon sum rank test, is performed to obtain P values and directions for all features on each node of the tree.</p>
<pre class="r"><code># a function to run the wilcox.test
test.func &lt;- function(X, Y) {  
  Y &lt;- as.numeric(factor(Y))
  obj &lt;- apply(X, 1, function(x) {                
    p.value &lt;- suppressWarnings(wilcox.test(x ~ Y)$p.value)
    p.value[is.na(p.value)] &lt;- 1
    e.sign &lt;- sign(mean(x[Y == 2]) - mean(x[Y == 1]))
    c(p.value, e.sign)          
  })
  df &lt;- data.frame(node = rowLinks(tse)$nodeNum,
                   pvalue = obj[1, ],
                   sign = obj[2, ])
  return(df)
}</code></pre>
<p><code>outW</code> is a list of 10 data frame, each of which is the differential analysis result of a feature.</p>
<pre class="r"><code># run differential analysis for all features
Y &lt;- colData(tse)$group
XList &lt;- assays(tse)

outW &lt;- lapply(1:10, FUN = function(i){
  df &lt;- test.func(X = XList[[i]], Y = Y)
  df$feature &lt;- names(XList)[i]
  return(df)
  })
names(outW) &lt;- names(XList)
head(outW$feature1)</code></pre>
<pre><code>        node    pvalue sign  feature
alias_1    1 0.6742060    1 feature1
alias_2    2 0.6327577   -1 feature1
alias_3    3 0.5603110   -1 feature1
alias_4    4 0.2436940   -1 feature1
alias_5    5 0.4274949    1 feature1
alias_6    6 0.2359014    1 feature1</code></pre>
</div>
<div id="get-candidates" class="section level1">
<h1>Get candidates</h1>
<p>Candidates are proposed separately for each feature.</p>
<pre class="r"><code>cand &lt;- lapply(outW, FUN = function(x) {
  getCand(tree = tree, score_data = x, node_column = &quot;node&quot;,
          p_column = &quot;pvalue&quot;, sign_column = &quot;sign&quot;, message = FALSE) })

# 10 features
names(cand)</code></pre>
<pre><code> [1] &quot;feature1&quot;  &quot;feature2&quot;  &quot;feature3&quot;  &quot;feature4&quot;  &quot;feature5&quot;  &quot;feature6&quot; 
 [7] &quot;feature7&quot;  &quot;feature8&quot;  &quot;feature9&quot;  &quot;feature10&quot;</code></pre>
<pre class="r"><code># candidates for feature 1 at specific t values
names(cand$feature1$candidate_list)</code></pre>
<pre><code> [1] &quot;0&quot;    &quot;0.01&quot; &quot;0.02&quot; &quot;0.03&quot; &quot;0.04&quot; &quot;0.05&quot; &quot;0.1&quot;  &quot;0.15&quot; &quot;0.2&quot;  &quot;0.25&quot;
[11] &quot;0.3&quot;  &quot;0.35&quot; &quot;0.4&quot;  &quot;0.45&quot; &quot;0.5&quot;  &quot;0.55&quot; &quot;0.6&quot;  &quot;0.65&quot; &quot;0.7&quot;  &quot;0.75&quot;
[21] &quot;0.8&quot;  &quot;0.85&quot; &quot;0.9&quot;  &quot;0.95&quot; &quot;1&quot;   </code></pre>
<p>Candidate proposed for different features are different at the same <code>t</code> if the signal patterns of features differ. The candidate proposed for features 6-10 that have no signal is the leaf level (the tree has 100 leaves).</p>
<pre class="r"><code># The number of nodes in a candidate generated at t = 0.3 for different features
sapply(cand, FUN = function(x){length(x$candidate_list[[&quot;0.3&quot;]])})</code></pre>
<pre><code> feature1  feature2  feature3  feature4  feature5  feature6  feature7  feature8 
       91        92        99        94        99       100       100       100 
 feature9 feature10 
      100       100 </code></pre>
</div>
<div id="evaluate-candidates" class="section level1">
<h1>Evaluate candidates</h1>
<pre class="r"><code>cand_list &lt;- lapply(cand, FUN = function(x){x$candidate_list})
res &lt;- evalCand(tree = tree, type = &quot;multiple&quot;,  
                levels = cand_list, score_data = outW,
                node_column = &quot;node&quot;, p_column = &quot;pvalue&quot;, sign_column = &quot;sign&quot;, 
                feature_column = &quot;feature&quot;)</code></pre>
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<p>In the simulation step, signals were created on different internal nodes for different features (<code>nd</code>). Signal nodes identified by <code>treeclimbR</code> are in <code>found</code>. Results are the numeric ID of signal nodes. <code>treeclimbR</code> has perfectly identified those signal nodes for features 1-5 in this toy example.</p>
<pre class="r"><code>out &lt;- res$output

# results of treeclimbR
(found &lt;- setNames(out[out$signal.node, ]$node,
                   out[out$signal.node, ]$feature))</code></pre>
<pre><code>feature1 feature2 feature3 feature4 feature5 
     147      186      185      118      132 </code></pre>
<pre class="r"><code># the truth
nd</code></pre>
<pre><code>feature1 feature2 feature3 feature4 feature5 
     147      186      185      118      132 </code></pre>
</div>
<div id="more-about-the-results" class="section level1">
<h1>More about the results</h1>
<p>There are 7 elements in the result list <code>res</code>.</p>
<pre class="r"><code>names(res)</code></pre>
<pre><code>[1] &quot;candidate_best&quot; &quot;output&quot;         &quot;candidate_list&quot; &quot;level_info&quot;    
[5] &quot;FDR&quot;            &quot;method&quot;         &quot;column_info&quot;   </code></pre>
<p>The best candidate levels picked for the first five features are different because we have assigned signals on different nodes. There are no signal for the last five features, so their best candidate levels are all on the leaf level (node ID: 1-100).</p>
<pre class="r"><code>res$candidate_best</code></pre>
<pre><code>$feature1
 [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19
[20]  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36  37  38
[39]  39  40  41  42  43  44  55  56  57  58  59  60  61  62  63  64  65  66  67
[58]  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86
[77]  87  88  89  90  91  92  93  94  95  96  97  98  99 100 147

$feature2
 [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19
[20]  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36  37  38
[39]  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54  55  56  57
[58]  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76
[77]  77  78  79  80  81  82  83  84  93  94  95  96  97  98  99 100 186

$feature3
 [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19
[20]  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36  37  38
[39]  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54  55  56  57
[58]  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76
[77]  77  78  79  80  81  82  85  86  87  88  89  90  91  92  93  94  95  96  97
[96]  98  99 100 185

$feature4
 [1]   1   2   3   4   5   6   7   8   9  10  11  12  19  20  21  22  23  24  25
[20]  26  27  28  29  30  31  32  33  34  35  36  37  38  39  40  41  42  43  44
[39]  45  46  47  48  49  50  51  52  53  54  55  56  57  58  59  60  61  62  63
[58]  64  65  66  67  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82
[77]  83  84  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99 100 118

$feature5
 [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19
[20]  20  21  22  23  24  25  26  27  30  31  32  33  34  35  36  37  38  39  40
[39]  41  42  43  44  45  46  47  48  49  50  51  52  53  54  55  56  57  58  59
[58]  60  61  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76  77  78
[77]  79  80  81  82  83  84  85  86  87  88  89  90  91  92  93  94  95  96  97
[96]  98  99 100 132

$feature6
  [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
 [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
 [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
 [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
 [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
 [91]  91  92  93  94  95  96  97  98  99 100

$feature7
  [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
 [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
 [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
 [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
 [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
 [91]  91  92  93  94  95  96  97  98  99 100

$feature8
  [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
 [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
 [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
 [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
 [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
 [91]  91  92  93  94  95  96  97  98  99 100

$feature9
  [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
 [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
 [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
 [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
 [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
 [91]  91  92  93  94  95  96  97  98  99 100

$feature10
  [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
 [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
 [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
 [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
 [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
 [91]  91  92  93  94  95  96  97  98  99 100</code></pre>
<p>There are two options <code>type = single</code> and <code>type = multiple</code> for <code>evalCand</code>. When <code>type = multiple</code> is used, the output table has one more column <code>feature</code>.</p>
<pre class="r"><code>out &lt;- res$output
head(out)</code></pre>
<pre><code>   node    pvalue sign  feature     adj.p signal.node
1:    1 0.6742060    1 feature1 0.9909277       FALSE
2:    2 0.6327577   -1 feature1 0.9909277       FALSE
3:    3 0.5603110   -1 feature1 0.9909277       FALSE
4:    4 0.2436940   -1 feature1 0.9909277       FALSE
5:    5 0.4274949    1 feature1 0.9909277       FALSE
6:    6 0.2359014    1 feature1 0.9909277       FALSE</code></pre>
<p>Signal nodes are found for different features.</p>
<pre class="r"><code>out[out$signal.node, ]</code></pre>
<pre><code>   node       pvalue sign  feature        adj.p signal.node
1:  147 3.383339e-06   -1 feature1 0.0006611045        TRUE
2:  186 3.357998e-06   -1 feature2 0.0006611045        TRUE
3:  185 3.341192e-06   -1 feature3 0.0006611045        TRUE
4:  118 3.349586e-06   -1 feature4 0.0006611045        TRUE
5:  132 3.282914e-06   -1 feature5 0.0006611045        TRUE</code></pre>
<p>Candidates proposed for different features at different <code>t</code> values are available in the <code>candidate_list</code></p>
<pre class="r"><code>names(res$candidate_list)</code></pre>
<pre><code> [1] &quot;feature1&quot;  &quot;feature2&quot;  &quot;feature3&quot;  &quot;feature4&quot;  &quot;feature5&quot;  &quot;feature6&quot; 
 [7] &quot;feature7&quot;  &quot;feature8&quot;  &quot;feature9&quot;  &quot;feature10&quot;</code></pre>
<pre class="r"><code>names(res$candidate_list$feature1)</code></pre>
<pre><code> [1] &quot;0&quot;    &quot;0.01&quot; &quot;0.02&quot; &quot;0.03&quot; &quot;0.04&quot; &quot;0.05&quot; &quot;0.1&quot;  &quot;0.15&quot; &quot;0.2&quot;  &quot;0.25&quot;
[11] &quot;0.3&quot;  &quot;0.35&quot; &quot;0.4&quot;  &quot;0.45&quot; &quot;0.5&quot;  &quot;0.55&quot; &quot;0.6&quot;  &quot;0.65&quot; &quot;0.7&quot;  &quot;0.75&quot;
[21] &quot;0.8&quot;  &quot;0.85&quot; &quot;0.9&quot;  &quot;0.95&quot; &quot;1&quot;   </code></pre>
<pre class="r"><code>res$FDR</code></pre>
<pre><code>[1] 0.05</code></pre>
<pre class="r"><code>res$method</code></pre>
<pre><code>[1] &quot;BH&quot;</code></pre>
<pre class="r"><code>res$column_info</code></pre>
<pre><code>$node_column
[1] &quot;node&quot;

$p_column
[1] &quot;pvalue&quot;

$sign_column
[1] &quot;sign&quot;

$feature_column
[1] &quot;feature&quot;</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.0 (2021-05-18)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] ggplot2_3.3.5                  ggnewscale_0.4.5              
 [3] TreeHeatmap_0.1.0              treeclimbR_0.1.4              
 [5] ape_5.5                        ggtree_3.0.4                  
 [7] TreeSummarizedExperiment_2.1.3 Biostrings_2.60.2             
 [9] XVector_0.32.0                 SingleCellExperiment_1.14.1   
[11] SummarizedExperiment_1.22.0    Biobase_2.52.0                
[13] GenomicRanges_1.44.0           GenomeInfoDb_1.28.4           
[15] IRanges_2.26.0                 S4Vectors_0.30.0              
[17] BiocGenerics_0.38.0            MatrixGenerics_1.4.3          
[19] matrixStats_0.60.1            

loaded via a namespace (and not attached):
  [1] utf8_1.2.2                  tidyselect_1.1.1           
  [3] lme4_1.1-27.1               grid_4.1.0                 
  [5] BiocParallel_1.26.2         Rtsne_0.15                 
  [7] aws.signature_0.6.0         flowCore_2.4.0             
  [9] munsell_0.5.0               codetools_0.2-18           
 [11] withr_2.4.2                 colorspace_2.0-2           
 [13] highr_0.9                   dirmult_0.1.3-4            
 [15] knitr_1.34                  ggsignif_0.6.3             
 [17] labeling_0.4.2              git2r_0.28.0               
 [19] GenomeInfoDbData_1.2.6      polyclip_1.10-0            
 [21] farver_2.1.0                pheatmap_1.0.12            
 [23] flowWorkspace_4.4.0         rprojroot_2.0.2            
 [25] vctrs_0.3.8                 treeio_1.16.2              
 [27] generics_0.1.0              TH.data_1.0-10             
 [29] xfun_0.26                   R6_2.5.1                   
 [31] doParallel_1.0.16           clue_0.3-59                
 [33] locfit_1.5-9.4              bitops_1.0-7               
 [35] gridGraphics_0.5-1          DelayedArray_0.18.0        
 [37] promises_1.2.0.1            scales_1.1.1               
 [39] multcomp_1.4-17             gtable_0.3.0               
 [41] Cairo_1.5-12.2              RProtoBufLib_2.4.0         
 [43] sandwich_3.0-1              workflowr_1.6.2            
 [45] rlang_0.4.11                GlobalOptions_0.1.2        
 [47] splines_4.1.0               rstatix_0.7.0              
 [49] lazyeval_0.2.2              hexbin_1.28.2              
 [51] broom_0.7.9                 reshape2_1.4.4             
 [53] yaml_2.2.1                  abind_1.4-5                
 [55] backports_1.2.1             httpuv_1.6.2               
 [57] RBGL_1.68.0                 tools_4.1.0                
 [59] ggplotify_0.1.0             ellipsis_0.3.2             
 [61] jquerylib_0.1.4             RColorBrewer_1.1-2         
 [63] ggdendro_0.1.22             Rcpp_1.0.7                 
 [65] plyr_1.8.6                  base64enc_0.1-3            
 [67] zlibbioc_1.38.0             purrr_0.3.4                
 [69] RCurl_1.98-1.5              FlowSOM_2.0.0              
 [71] ggpubr_0.4.0                viridis_0.6.1              
 [73] GetoptLong_1.0.5            zoo_1.8-9                  
 [75] haven_2.4.3                 ggrepel_0.9.1              
 [77] cluster_2.1.2               colorRamps_2.3             
 [79] fs_1.5.0                    magrittr_2.0.1             
 [81] ncdfFlow_2.38.0             data.table_1.14.0          
 [83] scattermore_0.7             openxlsx_4.2.4             
 [85] circlize_0.4.13             mvtnorm_1.1-2              
 [87] hms_1.1.0                   patchwork_1.1.1            
 [89] evaluate_0.14               XML_3.99-0.8               
 [91] rio_0.5.27                  jpeg_0.1-9                 
 [93] readxl_1.3.1                gridExtra_2.3              
 [95] shape_1.4.6                 ggcyto_1.20.0              
 [97] compiler_4.1.0              tibble_3.1.4               
 [99] crayon_1.4.1                ggpointdensity_0.1.0       
[101] minqa_1.2.4                 htmltools_0.5.2            
[103] ggfun_0.0.3                 later_1.3.0                
[105] tidyr_1.1.3                 aplot_0.1.0                
[107] RcppParallel_5.1.4          aws.s3_0.3.21              
[109] tweenr_1.0.2                ComplexHeatmap_2.8.0       
[111] MASS_7.3-54                 boot_1.3-28                
[113] Matrix_1.3-4                car_3.0-11                 
[115] diffcyt_1.12.0              igraph_1.2.6               
[117] forcats_0.5.1               pkgconfig_2.0.3            
[119] foreign_0.8-81              xml2_1.3.2                 
[121] foreach_1.5.1               bslib_0.2.5.1              
[123] yulab.utils_0.0.2           stringr_1.4.0              
[125] digest_0.6.27               ConsensusClusterPlus_1.56.0
[127] graph_1.70.0                rmarkdown_2.10             
[129] cellranger_1.1.0            tidytree_0.3.5             
[131] edgeR_3.34.1                curl_4.3.2                 
[133] rjson_0.2.20                nloptr_1.2.2.2             
[135] lifecycle_1.0.0             nlme_3.1-152               
[137] jsonlite_1.7.2              carData_3.0-4              
[139] viridisLite_0.4.0           limma_3.48.3               
[141] fansi_0.5.0                 pillar_1.6.2               
[143] lattice_0.20-44             fastmap_1.1.0              
[145] httr_1.4.2                  survival_3.2-13            
[147] glue_1.4.2                  zip_2.2.0                  
[149] png_0.1-7                   iterators_1.0.13           
[151] Rgraphviz_2.36.0            ggforce_0.3.3              
[153] stringi_1.7.4               sass_0.4.0                 
[155] CytoML_2.4.0                latticeExtra_0.6-29        
[157] dplyr_1.0.7                 cytolib_2.4.0              </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
